Upgrade to GraalVM 22.3.0

JaroslavTulach opened the pull request:
### Pull Request Description

Upgrading to GraalVM 22.3.0.

### Important Notes

- Removed all deprecated `FrameSlot`, and replaced them with frame indexes - integers.
  - Add more information to `AliasAnalysis` so that it also gathers these indexes.
- Add quick build mode option to `native-image` as default for non-release builds
- `graaljs` and `native-image` should now be downloaded via `gu` automatically, as dependencies.
- Remove `engine-runner-native` project - native image is now build straight from `engine-runner`.
  - We used to have `engine-runner-native` without `sqldf` in classpath as a workaround for an internal native image bug.
- Fixed chrome inspector integration, such that it shows values of local variables both for current stack frame and caller stack frames.
  - There are still many issues with the debugging in general, for example, when there is a polyglot value among local variables, a `NullPointerException` is thrown and no values are displayed.
- Removed some deprecated `native-image` options
- Remove some deprecated Truffle API method calls.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

Commit:
Upgrade to GraalVM 22.1.0 - mostly FrameSlot related fixes

JaroslavTulach commented on build.sbt:
```diff
@@ -15,7 +15,7 @@ import java.io.File
 
 val scalacVersion = "2.13.8"
-val graalVersion = "21.3.0"
+val graalVersion = "22.1.0"
```
GraalVM 22.2 is out - wouldn't it be better to update to newest version?

jtulach replied:
Yes, definitely. I want to ultimately get to 22.3.0 version, I just wanted to make this step by step, rather than updating straight to the newest version.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/codegen/IrToTruffle.scala:
```diff
@@ -326,12 +324,12 @@ class IrToTruffle(
               expressionProcessor.scope,
               moduleScope,
-              () => bodyBuilder.bodyNode(),
+              bodyBuilder.bodyNode(),
```
The `LazyBodyNode` has been introduced by #3429 and according to [the measurements](https://www.pivotaltracker.com/n/projects/2539304/stories/182024911) it speeds the IDE startup by 21%:

> 1.5s out of 7s speedup

I don't think we can give that up. We need to find a way to convert nodes lazily and make it work even with `int` based slots. It shouldn't be that hard - traversing `IR` shall give us the exact amount of needed slots without creating the Truffle nodes.

kustosz commented on engine/runtime/src/main/java/org/enso/interpreter/node/ClosureRootNode.java:
```diff
@@ -86,9 +87,9 @@ public Object execute(VirtualFrame frame) {
+    frame.setObject(this.getStateFrameSlotIdx(), state);
     Object result = body.executeGeneric(frame);
-    state = FrameUtil.getObjectSafe(frame, this.getStateFrameSlot());
+    state = frame.getObject(this.getStateFrameSlotIdx());
```
Seems like the proper replacement for `FrameUtil.getObjectSafe` should be `getObjectStatic`?

jtulach replied:
get/setObjectStatic uses static slots, and for such slots Truffle does not track FrameSlotKind, more specifically, there is only FrameSlotKind.Static. This means that for static slots, we would have to keep the information about its kind internally.

Whereas normal slots (that is indexed slots) conform to "standard lifecycle" - They begin with FrameSlotKind.Illegal, from which they go to, e.g., FrameSLotKind.Double and from there to FrameSlotKind.Object in which they stay - this is what we do, e.g., in AssignmentNode.

On the other hand, stateFrameSlotIdx is a special case - it conforms to monadic state slot which is implicitly assigned to every FrameDescriptor, moreover, it seems that in Enso, this will always be of FrameSlotKind.Object. So for this one particular slot, we could initialize it with setObjectStatic and access it via getObjectStatic, but I personally would prefer to keep it simple and use normal slots everywhere, unless we migrate to static slots altogether.

kustosz reviewed (commented):
I'm happy with this as long as it works and does not cause performance regressions :)

JaroslavTulach commented:
CI run [shows](https://github.com/enso-org/enso/runs/7967611554) that the `DebuggingEnsoTest` is failing. Local variables aren't visible as the default visualization doesn't work for `int` based variable indexes. That's going to take a time, but it is an opportunity to display variable values properly.

JaroslavTulach commented:
[Formatting CI check](https://github.com/enso-org/enso/runs/7967611255) seems to be using GraalVM 21.3 instead of 22.1. Ccing @mwu-tow ...

jtulach commented:
> CI run [shows](https://github.com/enso-org/enso/runs/7967611554) that the `DebuggingEnsoTest` is failing. Local variables aren't visible as the default visualization doesn't work for `int` based variable indexes. That's going to take a time, but it is an opportunity to display variable values properly.

Right. The reason why this particular test does not work is that the default implementation of NodeLibrary ( DefaultfNodeExports$DefaultScope.readMember ) reads only auxiliary slots from the frame, but ignores all the indexed slots. For, now deprecated, FrameSlot it was OK, because the default implementation just traversed all the FrameSlots. We have to correctly implement this behavior by exporting scope messages from NodeLibrary in BaseNode or ExpressionNode. The getScope message would return a TruffleObject representing a scope that would export all member and scope messages from InteropLibrary, where there are parent scopes, among other things.

I suggest we ignore, or provide a quick fix, for this test for now, and create a follow-up issue to implement variable displaying in the Truffle debugger properly.

mwu-tow commented:
@JaroslavTulach 

> [Formatting CI check](https://github.com/enso-org/enso/runs/7967611255) seems to be using GraalVM 21.3 instead of 22.1. Ccing @mwu-tow ...

The workflow has [hardcoded GraalVM version](https://github.com/enso-org/enso/blob/develop/.github/workflows/formatting.yml#L11), it needs to be bumped as well.

You don't need to do this for my build-script-based workflows, as the build script parses `build.sbt` to figure out proper GraalVM version.

Commit:
Some implementation of scopes for debugging purposes

JaroslavTulach committed:
Ignoring the stackoverflow test for now

JaroslavTulach committed:
Removing no longer existing import

JaroslavTulach committed:
Eliminate FrameScope reference

Commit:
Update formatting workflow to GraalVM 22.1.0

Commit:
Merge branch 'develop' into wip/enso/UpdateGraalVM22_182845176

Commit:
Get rid of createCallTarget deprecation

Commit:
Formating

Commit:
[WIP] Add FrameIndexAnalysis compiler pass with tests

Commit:
Add anonymous function tests

Commit:
Fix binding expression recursive analysis

Commit:
Add test for pass configuration

Commit:
LocalScope uses alias analysis for frame indexes

Commit:
Add allDefinitions test to AliasAnalysis

Commit:
Small fixes on FrameIndexAnalysis

Commit:
Remove FrameIndexAnalysis

Commit:
Cosmetics

Commit:
Revert to lazy body building

Commit:
Fix deepCopy of Scope.

Commit:
Merge branch 'develop' into wip/enso/UpdateGraalVM22_182845176

Commit:
Cosmetics

Commit:
[WIP] Ignore DebuggingEnsoTest

Commit:
Bump GraalVM version to 22.3.0

Commit:
Enable quick native image build if not in release

Commit:
Remove getCurrentFrame deprecation

Commit:
Add some docs

Commit:
Add truffle boundaries to some toDisplayString methods.

native image build now fails with "Encoding two deopt entries in the same bci", and this seems like the easiest fix.

Commit:
native image fix - Use createSystemThread in SerializationManager

Fix of what seems to be a bug in native image.

Commit:
Merge branch 'develop' into wip/enso/UpdateGraalVM22_182845176

Commit:
Remove deprecated native-image options

Commit:
Remove runner-native project.

It is not required anymore.

Commit:
Add installNativeImage and installGraalJs build tasks

Commit:
Bump GraalVM version in other files

Commit:
Remove inner context builder deprecation

Commit:
Cosmetics

Commit:
Merge branch 'develop' into wip/enso/UpdateGraalVM22_182845176

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/node/ExpressionNode.java:
```diff
@@ -85,6 +90,16 @@ public void setId(UUID id) {
 
+  @ExportMessage
+  public boolean hasScope(Frame frame) {
+    throw new UnsupportedOperationException("unimplemented");
```
I have created https://www.pivotaltracker.com/story/show/183798184 to track this

Akirathan commented on engine/runtime/src/test/java/org/enso/interpreter/test/DebuggingEnsoTest.java:
```diff
@@ -18,10 +18,12 @@
 
 public class DebuggingEnsoTest {
   @Test
+  @Ignore
```
Tracked by https://www.pivotaltracker.com/story/show/183798184

JaroslavTulach replied:
OK to temporarily disable. However @jdunkerley & others are using `enso --inspect` option and right now it not going to show values of local variables - e.g. it's usefulness is going to be seriously impacted.

Akirathan replied:
Done.

Akirathan commented on project/NativeImage.scala:
```diff
@@ -110,15 +110,14 @@ object NativeImage {
         Seq("--no-fallback", "--no-server") ++
-        initializeAtBuildtimeOptions ++
-        memoryLimitOptions ++ initializeAtRuntimeOptions ++
+        Seq("--initialize-at-build-time=") ++
```
Tracked by https://www.pivotaltracker.com/story/show/183794559

Akirathan commented on build.sbt:
```diff
@@ -1602,20 +1644,28 @@ lazy val `engine-runner-native` = project
+          "io.methvin.watchservice.jna.CarbonAPI",
         )
       )
+      .dependsOn(installNativeImage)
```
@hubertp Multiple tasks now depend on `installNativeImage` task. `installNativeImage` should be *synchronized*, i.e., should be called at most once from just one thread. Is there any way how to achieve that in `sbt`? A quick workaround could be to introduce a mutex, but that seems ugly.

hubertp replied:
Added a commit that solves this, I think

Akirathan replied:
Thanks, I will do the same for `installGuJS`

Commit:
Cosmetics

Commit:
Update changelog

Commit:
Merge branch 'develop' into wip/enso/UpdateGraalVM22_182845176

Commit:
Add TruffleString specialization to CoercePrimitiveNode

Commit:
Bump Java version constants in Http_Spec.enso

JaroslavTulach commented:
Checking the failures is see two:
- [correctly pattern match on array like values](https://github.com/enso-org/enso/actions/runs/3472406583/jobs/5803192679#step:10:6920)
- [ArrayList doesn't equal ArrayList](https://github.com/enso-org/enso/actions/runs/3472406583/jobs/5803192679#step:10:7203)

Great job! Looks like you'll be able to merge soon.

hubertp commented on docs/CONTRIBUTING.md:
````diff
@@ -529,6 +529,12 @@ enso --run test/Geo_Tests
+Or to run just a single test (e.g., `Duration_Spec.enso`):
+
+```bash
+enso --no-ir-caches --in-project test/Tests --run test/Tests/src/Data/Time/Duration_Spec.enso
````
The previous command doesn't mention caches either
```suggestion
enso --in-project test/Tests --run test/Tests/src/Data/Time/Duration_Spec.enso
```

hubertp committed:
Use .result to avoid multiple task executions

Removed spurious invocations.

hubertp commented on engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/AliasAnalysis.scala:
```diff
@@ -414,9 +415,10 @@ case object AliasAnalysis extends IRPass {
+        val definition =
           Occurrence.Def(labelId, label.name, label.getId, label.getExternalId)
-        )
+        parentScope.add(definition)
```
Since this is repeated, maybe worth having `addWithDefintion` method in `Scope` that does `add` and `addDefinition`?

Akirathan replied:
Probably. But there is also a call site that calls only `addDefinition` and not `add`, which is vital. So instead of having `add`, `addDefinition` and `addWithDefinition`, I will rather repeat `add` and `addDefinition` on few places.

hubertp commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/scope/LocalScope.scala:
```diff
@@ -108,6 +144,48 @@ class LocalScope(
+  }
+
+  /** Gather local variables from the alias scope information.
+    * Does not include any vairables from the parent scopes.
```
```suggestion
    * Does not include any variables from the parent scopes.
```

hubertp commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/scope/LocalScope.scala:
```diff
@@ -108,6 +144,48 @@ class LocalScope(
+
+  /** Gather local variables from the alias scope information.
+    * Does not include any vairables from the parent scopes.
+    * @return
```
?

hubertp commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/scope/LocalScope.scala:
```diff
@@ -66,38 +81,59 @@ class LocalScope(
+    * @return
+    */
+  def monadicStateSlotIdx: Int = {
+    for {
+      (_, name) <- internalSlots
+      idx       <- internalSlots.indices
+    } {
+      if (name == "<<monadic_state>>") {
+        return idx
+      }
+    }
+    throw new IllegalStateException(
+      "<<monadic_state>> slot should be present in every frame descriptor"
+    )
```
```suggestion
    internalSlots.zipWithIndex
      .find { case ((_, name), _) => name == "<<monadic_state>>" }
      .map(_._2)
      .getOrElse(
        throw new IllegalStateException(
          "<<monadic_state>> slot should be present in every frame descriptor"
        )
      )
```

hubertp commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/scope/LocalScope.scala:
```diff
@@ -28,20 +28,35 @@ import scala.jdk.CollectionConverters._
+    * Every tuple of the list denotes frame slot kind and its name.
+    * Note that `info` for a frame slot is not used by Enso.
+    */
+  private val internalSlots: List[(FrameSlotKind, String)] = List(
```
It feels like this is constant and should be defined in a companion object of `LocalScope`

hubertp reviewed (approved):
Some suggestions but LGTM otherwise. Also, slightly changed sbt to avoid multiple installations of `native-image`

Akirathan committed:
Apply suggestions from code review

Co-authored-by: Hubert Plociniczak <hubert.plociniczak@gmail.com>

Akirathan committed:
Remove cache opt from CONTRIBUTING.md

Co-authored-by: Hubert Plociniczak <hubert.plociniczak@gmail.com>

Commit:
[WIP] Add basic NodeLibrary support

Commit:
Merge branch 'wip/enso/UpdateGraalVM22_182845176' of github.com:enso-org/enso into wip/enso/UpdateGraalVM22_182845176

Commit:
Implement NodeLibrary support.

- Display local variables from current method.
- Display local variables of callers.

Commit:
Small refactorings in LocalScope.

Commit:
Merge branch 'develop' into wip/enso/UpdateGraalVM22_182845176

Commit:
Add specialization for meta object to IsSameObjectNode

Commit:
Merge branch 'develop' into wip/enso/UpdateGraalVM22_182845176

4e6 commented on engine/runtime/src/main/scala/org/enso/interpreter/runtime/scope/LocalScope.scala:
```diff
@@ -66,38 +74,56 @@ class LocalScope(
+    */
+  def monadicStateSlotIdx: Int = {
+    internalSlots.zipWithIndex
+      .find { case ((_, name), _) => name == "<<monadic_state>>" }
```
"<monadic_state>" string could be a constant

4e6 commented on project/DistributionPackage.scala:
```diff
@@ -506,13 +506,14 @@ object DistributionPackage {
       * @param os the system type
       * @param graalDir the directory with a GraalVM distribution
       * @param arguments the command arguments
+     *                  @return Stdout from the `gu` command.
```
format

4e6 reviewed (approved):
I seems that each GraalVM update is becoming increasingly harder. Happy that it is finally done :heart:

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/node/ExpressionNode.java:
```diff
@@ -63,7 +79,11 @@ public void setSourceLocation(int sourceStartIndex, int sourceLength) {
   @Override
   public SourceSection getSourceSection() {
-    return EnsoRootNode.findSourceSection(getRootNode(), sourceStartIndex, sourceLength);
+    if (this instanceof ExpressionNodeWrapper wrapper) {
```
there is a `isWrapper` method that does this instance check and it seems that in `unwrapDelegate` we don't mind casting the node so maybe stick to one style?

Akirathan replied:
Probably. But then I don't get this fancy Java pattern matching instanceof. Nevertheless, I would not bother with this code that much, as it will probably be rewritten within a few weeks, once I get to a more proper support for debugging.

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/DebugLocalScope.java:
```diff
@@ -0,0 +1,278 @@
+  private final EnsoRootNode rootNode;
+  /** All the bindings, including the parent scopes. */
+  private final Map<String, FramePointer> allBindings;
+  /**
```
style nit: some newlines are missing between declarations

Akirathan replied:
Thanks for noticing. I thought that `sbt javafmtAll` would take care of this for me.

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/DebugLocalScope.java:
```diff
@@ -0,0 +1,278 @@
+    this.rootNode = rootNode;
+    this.frame = frame;
+    this.allBindings = rootNode.getLocalScope().flattenBindings();
+    this.bindingsByLevels =
```
nit: it feels like this logic should be moved to the aux constructor rather than using `null`
At least that way I wouldn't have to look at both constructors to understand what is going on.

Akirathan replied:
Sounds good. Done

JaroslavTulach committed:
Skip 'deep' gu location, if it doesn't exist

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/DebugLocalScope.java:
```diff
@@ -0,0 +1,278 @@
+      if (framePointer.getParentLevel() > maxParentLevel) {
+        maxParentLevel = framePointer.getParentLevel();
+      }
+    }
```
That feels like it could be just rewritten with `java.util.Collections.max`

Commit:
Fix some review comments

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/DebugLocalScope.java:
```diff
@@ -0,0 +1,278 @@
+    List<String> members = new ArrayList<>();
+    for (int i = bindingsByLevelsIdx; i < bindingsByLevels.size(); i++) {
+      members.addAll(bindingsByLevels.get(i));
+    }
```
`bindingsByLevels.forEach`?

Commit:
Merge branch 'wip/enso/UpdateGraalVM22_182845176' of github.com:enso-org/enso into wip/enso/UpdateGraalVM22_182845176

hubertp reviewed (approved):
nits

mwu-tow commented on build.sbt:
```diff
@@ -1303,6 +1305,79 @@ lazy val `runtime-language-epb` =
+  }
+}
+
+lazy val installNativeImage = taskKey[Unit]("Install native-image GraalVM component")
```
This logic already is present in the RRust build script, if you use it to enter `sbt` shell then you'll have native-image installed.

Akirathan replied:
I have just recently noticed that. Nevertheless we still need `installGraalJs`, as it is not shipped by default in the latest GraalVM release. Moreover, this task does not install anything, if `native-image` is already installed. The same is true for `installGraalJs`.  Besides, I don't know how other engine devs, but I don't usually use the rust build script, I just use `sbt`, which makes these two tasks very convenient.

Commit:
Make some methods in DebugLocalScope more functional

Commit:
Merge branch 'develop' into wip/enso/UpdateGraalVM22_182845176
